src/Chapter3.hs

```diff
@@ -1006,7 +1122,78 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
-
+--Assuming data for each player (monster/knight)
+newtype PlayerHealth = PlayerHealth Int
+newtype PlayerAttack = PlayerAttack Int
+newtype PlayerDefense = PlayerDefense Int
+
+--All possible actions
+data BattleAction = Attack | DrinkPotion Int | CastSpell Int | MonsterAttack | MonsterRun
+
+data BattleResult = Round FighterData MonsterData | Victory | Defeat
+data FighterData = FighterData
+  {
+    fighterHealth :: PlayerHealth
+  , fighterAttack :: PlayerAttack
+  , fighterDefense :: PlayerDefense
+  , fighterActions :: [BattleAction]
+  }
+
+data MonsterData = MonsterData
+  {
+    monsterHealth :: PlayerHealth
+  , monsterAttack :: PlayerAttack
+  , monsterActions :: [BattleAction]
+  }
+
+--Attacking class
+class PlayerAttacking a where
+  getHealth :: a -> PlayerHealth
+  getAttack :: a -> PlayerAttack
+
+instance PlayerAttacking FighterAttack where
+  getHealth = fighterHealth
+  getAttack = fighterAttack
+
+instance PlayerAttacking MonsterAttack where
+  getHealth = monsterHealth
+  getAttack = monsterAttack
+
+--Defensing
+class PlayerDefensing a where
+  getDefense :: a -> PlayerDefense
+
+instance PlayerDefensing FighterDefense where
+  getDefense = fighterDefense
+
+roundFighting :: BattleAction -> FighterAttack -> MonsterAttack -> BattleResult
```

Comment:
It looks like data types `FitherData` and `MonsterData` are not used in this function :thinking: 
Also, this function works with specific types, not polymorphic. Probably you need to use typeclasses defined before